language/resource-api/README.md

```diff
@@ -355,6 +363,144 @@ For example, in the Puppet runtime environment this is bound to the `strict` set
+- If `insync?` returns `true`, the property is treated as insync.
+- If `insync?` returns `false`, the property is treated as out of sync and will cause the default change reporting.
+  - e.g. `Notice: /Stage[main]/Main/File[/tmp/foo]/mode: mode changed <is value> to <should value>`
+- If `insync?` returns a string, the property is treated as out of sync and the string is used for change reporting instead of the default. This can be used to provide additional information about the reason for a change.
+  - In the `test_custom_insync` provider example above, you would get the message `Notice: /Stage[main]/Main/Test_custom_insync[something]/some_array: Adding missing members ['SomeValue']` if `SomeValue` was specified as a member of `some_array` but not found in the actual state returned from `get`.
+  - This enables more accurate and useful change reporting than the default is-to-should reporting surfaced by Puppet.
+
+The `insync?(context, name, property_name, is_hash, should_hash)` method receives the following arguments:
+
+- a `context` to access the runtime environment,
+- the `name` of the resource being checked,
+- the `property_name` that is currently being inspected for insync status,
+- the current state of the resource in `is_hash`, and
+- the desired state (canonicalized) in `should_hash`.
+
+For resources which specify the `custom_insync` feature flag but have _no_ defined properties the Resource API adds a spectial hidden property, `rsapi_custom_insync_trigger`, which ensures that `insync?` is still called for the resource.
+This hidden property is not used or seen anywhere else in the provider, only in the `insync?` method.
+
+As with all calls to `insync?`, a custom change reporting message can be surfaced by returning a string instead of `true` or `false`.
+If no custom change reporting message is returned as a string, the change reporting in Puppet (if `false` is returned) will always be:
+
+> Custom insync logic determined that this resource is out of sync
+
+```ruby
+# lib/puppet/type/test_custom_insync_hidden_property.rb
+Puppet::ResourceApi.register_type(
+  name: 'test_custom_insync_hidden_property',
+  docs: <<-EOS,
+      This type provides Puppet with the capabilities to manage ...
+    EOS
+  features: ['custom_insync'],
+  attributes:   {
+    name:        {
+      type:      'String',
+      desc:      'The name of the resource you want to manage.',
+      behaviour: :namevar,
+    },
+    force: {
+      type:      'Boolean',
+      desc:      'If true, the resource will be treated as not being insync.',
+      behaviour: :parameter,
+      default:   false,
+    },
+  },
+)
+
+# lib/puppet/provider/test_custom_insync_hidden_property/test_custom_insync_hidden_property.rb
+class Puppet::Provider::TestCustomInsyncHiddenProperty::TestCustomInsyncHiddenProperty
```

Comment:
What is this example showing compared to the first one?

Reply:
This shows an example _without_ an insyncable property - which adds a hidden property to ensure that `insync?` is triggered.

Reply:
Ah okay, maybe make that clearer?

Reply:
I've moved this entire section into a subheading, does that address this concern?

https://github.com/puppetlabs/puppet-specifications/pull/153/files#diff-23f362101839bf8ed43d75a53178cf3cb35aa2d8169fc0163f3a7619f87ef79fR457

Reply:
Yes that is better, thanks!